cupyx/scipy/ndimage/filters.py

```diff
@@ -0,0 +1,195 @@
+
+
+def convolve(input, weights, output=None, mode='reflect', cval=0.0, origin=0):
+    """Multi-dimensional convolution.
+
+    The array is convolved with the given kernel.
+
+    Args:
+        input (cupy.ndarray): The input array.
+        weights (cupy.ndarray): Array of weights, same number of dimensions as
+            input
+        output (cupy.ndarray or None): The array in which to place the output.
+        mode (str): The array borders are handled according to the given mode
+            (``'reflect'``, ``'constant'``, ``'nearest'``, ``'mirror'``,
+            ``'wrap'``). Default is ``'reflect'``.
+        cval (scalar): Value to fill past edges of input if mode is
+            ``constant``. Default is ``0.0``.
+        origin (scalar or tuple of scalar): The origin parameter controls the
+            placement of the filter, relative to the center of the current
+            element of the input. Default of 0 is equivalent to
+            ``(0,)*input.ndim``.
+
+    Returns:
+        cupy.ndarray: The result of convolution.
+
+    .. seealso:: :func:`scipy.ndimage.convolve`
+    """
+    return _correlate_or_convolve(input, weights, output, mode, cval, origin,
+                                  True)
+
+
+def _get_output(output, input, shape=None):
+    if shape is None:
+        shape = input.shape
+    if isinstance(output, cupy.ndarray):
+        if output.shape != tuple(shape):
+            raise ValueError('output shape is not correct')
+    else:
+        dtype = output
+        if dtype is None:
+            dtype = input.dtype
+        output = cupy.zeros(shape, dtype)
+    return output
+
+
+def _correlate_or_convolve(input, weights, output, mode, cval, origin,
+                           convolution):
+    if input.dtype in (numpy.complex64, numpy.complex128, numpy.complex256):
+        raise TypeError('Complex type not supported.')
+    if not hasattr(origin, '__getitem__'):
+        origin = [origin, ] * input.ndim
+    else:
+        origin = list(origin)
+    wshape = [ii for ii in weights.shape if ii > 0]
+    if len(wshape) != input.ndim:
+        raise RuntimeError('filter weights array has incorrect shape.')
+    if convolution:
+        weights = weights[tuple([slice(None, None, -1)] * weights.ndim)]
+        for ii in range(len(origin)):
+            origin[ii] = - origin[ii]
+            if not weights.shape[ii] & 1:
+                origin[ii] -= 1
+    for _origin, lenw in zip(origin, wshape):
+        if (lenw // 2 + _origin < 0) or (lenw // 2 + _origin > lenw):
+            raise ValueError('invalid origin')
+    if mode not in ('reflect', 'constant', 'nearest', 'mirror', 'wrap'):
+        msg = 'boundary mode not supported (actual: {}).'.format(mode)
+        raise RuntimeError(msg)
+
+    output = _get_output(output, input)
+    if weights.size == 0:
+        return output
+    input = cupy.ascontiguousarray(input)
+    weights = cupy.ascontiguousarray(weights, cupy.float64)
```

Comment:
Why `float64`?

Reply:
That is because SciPy implementation always uses `float64` for weights. 
https://github.com/scipy/scipy/blob/e1e44d12637997606b1bcc0c6de232349e11eee0/scipy/ndimage/filters.py#L602
My first implementation did not use `float64` here hence causing some output incompatibility to SciPy implementation as @grlee77 indicated.

Reply:
https://github.com/cupy/cupy/pull/2483#issuecomment-533690790

Reply:
Thank you for the reference.
Could you write a comment? It's not directly clear in the code.

(Btw it might not be optimal to use float64 in GPUs. I will make an issue later for the policy whether to weigh on performance or precision.)

Reply:
If the GPU has full `float64` capability (like Tesla V100), there is no performance problem with using `float64`, but if it does not, there might be a performance issue...
One solution without harming output compatibility is to use `float32` when, for example, input dtype is either `int8`, `int16` or `float32`. What would you think it is better?

Reply:
Even if `input.dtype` is `float32`, there will be some degradation in precision if accumulation is done in `float32`.

I think it cannot be easily decided in this PR, because it's about the policy throughout CuPy.
So how about proceeding with current implementation and discuss later in another issue?

Reply:
OK, then I will revert the fix about weights dtype and just add comments on this.

Reply:
Thanks, please write a comment mentioning scipy.

Reply:
The comment about compatibility with scipy added.

Reply:
Issue about the policy: https://github.com/cupy/cupy/issues/2505